### PR TITLE
[Gluten-531] Add substrait spark module

### PR DESCRIPTION
**Pull request by baibaichen**
## What changes were proposed in this pull request?

This is the first PR of https://github.com/oap-project/gluten/issues/531 for using [substrait-java](https://github.com/substrait-io/substrait-java), it adds substrait spark module(refer to [isthmus](https://github.com/substrait-io/substrait-java/tree/main/isthmus)), and using a new way to convert spark expression to substriat proto expression.

To support a new expression, we needn't inherit from Expression, and just add a defition in FuntionsMapping, for example:

```scala
  val AGGREGATE_SIGS: Seq[Sig] = Seq(
    s[Sum]("sum"),
    s[Count]("count"),
    s[HyperLogLogPlusPlus]("approx_count_distinct")
  )
```

## How was this patch tested?
1. After optimizing, SQL is converted to spark `LogicalPlan`, and then converted substrait plan. 
2. The converted substrait plan is restored to spark logical plan, 
3. Converting logical plan to substrait plan again.
4. Comparing these two substrait plans, if they are equal, then converting is OK.

The codes are there:

``` scala
  def assertSqlSubstraitRelRoundTrip(query: String): LogicalPlan = {
    val logicalPlan = plan(query)
    val pojoRel = new SubstraitRelVisitor().visit(logicalPlan)
    val logicalPlan2 = new SubstraitLogicalPlanConverter(spark = spark).convert(pojoRel)
    val pojoRel2 = new SubstraitRelVisitor().visit(logicalPlan2)
    pojoRel2.shouldEqualPlainly(pojoRel)
    logicalPlan2
  }
```

**Currently, we can successfully convert TPCH SQLs**

## other
Using build-helper-maven-plugin to add different sources to support spark3.2 and spark 3.3, the reason I am doing this is that I want to move Subtrait-spark into subtrait-java in the future

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -310,6 +316,26 @@
           </exclusion>
         </exclusions>
       </dependency>
+      <dependency>
+        <groupId>com.google.guava</groupId>
+        <artifactId>guava</artifactId>
```
Comment (jinchengchenghh) [inline pom.xml:321]: Does this dependency scope should be runtime?

Comment (baibaichen) [reply pom.xml:321]: Thanks for pointing this

File: substrait/substrait-spark/src/main/resources/spark.yml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  -
+    name: unscaled
+    description:  >-
+      Return the unscaled Long value of a Decimal, assuming it fits in a Long.
+      Note: this expression is internal and created only by the optimizer, 
+      we don't need to do type check for it.
+    impls:
+      - args:
+          - value: DECIMAL<P,S>
+        return: i64
```
Comment (rui-mo) [inline substrait/substrait-spark/src/main/resources/spark.yml:20]: Does current logic rely on this yml file? Looks like TPC-H related functions are not fully added here.

File: substrait/substrait-spark/src/main/scala/io/substrait/spark/expression/FunctionConverter.scala [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+      arguments: Seq[FunctionArg],
+      outputType: Type): T
+  def getSigs: Seq[Sig]
+
+  private def init(
+      functions: Seq[F]): (ju.Map[Class[_], FunctionFinder[F, T]], Multimap[String, Sig]) = {
+    val alm = ArrayListMultimap.create[String, F]()
+    functions.foreach(f => alm.put(f.name().toLowerCase(ju.Locale.ROOT), f))
+
+    val calciteOperators = ArrayListMultimap.create[String, Sig]()
```
Comment (rui-mo) [inline substrait/substrait-spark/src/main/scala/io/substrait/spark/expression/FunctionConverter.scala:56]: Looks like `calciteOperators` is used for function signatures, how does it relate to operator?

File: substrait/substrait-spark/src/main/resources/spark.yml (hunk 2 above)
Comment (baibaichen) [reply substrait/substrait-spark/src/main/resources/spark.yml:20]: Those functions are included substrait-core

File: substrait/substrait-spark/src/main/scala/io/substrait/spark/expression/FunctionConverter.scala (hunk 3 above)
Comment (baibaichen) [reply substrait/substrait-spark/src/main/scala/io/substrait/spark/expression/FunctionConverter.scala:56]: This is Sig's definition:

`case class Sig(expClass: Class[_], name: String, builder: Seq[Expression] => Expression)`

and how we create one::

```scala
  private def s[T <: Expression: ClassTag](name: String): Sig = {
    val builder = FunctionRegistryBase.build[T](name, None)._2
    Sig(scala.reflect.classTag[T].runtimeClass, name, builder)
  }
```

Please node: there is a one-to-many relationship between the name and the expression, for example, `Add`   is replaced by `DateAdd`  by `Analyzer` if we add days to date

Comment (rui-mo) [reply substrait/substrait-spark/src/main/scala/io/substrait/spark/expression/FunctionConverter.scala]: Thanks for your explanation!

**Comment by zhixingheyi-tian** [comment]
Hi @baibaichen ,
Two  questions here:
-  Noticed this submodule didn't  put parent pom.xml, then how to build and test this patch locally?
-  Why not convert physical plan, any consideration?
Thanks，

File: substrait/substrait-spark/pom.xml [hunk 4]
```diff
@@ -0,0 +1,159 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
```
Comment (baibaichen) [inline substrait/substrait-spark/pom.xml:5]: @zhixingheyi-tian  did you mean this?

**Comment by baibaichen** [comment]
@zhixingheyi-tian 
> * Why not convert physical plan, any consideration?

We convert physical plan in gluten-core.  Logical plan converter is used for testing `ToSubstraitExpression`

File: substrait/substrait-spark/pom.xml (hunk 4 above)
Comment (zhixingheyi-tian) [reply substrait/substrait-spark/pom.xml:5]: I mean put into  here, https://github.com/oap-project/gluten/blob/9ed615443cb420d0682f07e7b7e70bfa00228569/pom.xml#L229

Comment (baibaichen) [reply substrait/substrait-spark/pom.xml:5]: gluten-core depends on this module, no need to put it in profiles

File: pom.xml [hunk 5]
```diff
@@ -32,6 +32,7 @@
   <modules>
     <module>shims</module>
     <module>gluten-core</module>
+    <module>substrait/substrait-spark</module>
```
Comment (baibaichen) [inline pom.xml:35]: it's here @zhixingheyi-tian

**Review by FelixYBW — APPROVED** [review]

File: pom.xml (hunk 5 above)
Comment (zhixingheyi-tian) [reply pom.xml:35]: Seen, Thanks

**Pull request merged**